pkg/validation/validation_test.go

```diff
@@ -0,0 +1,480 @@
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+		"spec.activeDeadlineSeconds: Invalid value: -42: must be greater than or equal to 0": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {},
+				},
+				ActiveDeadlineSeconds: &negative64,
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+		"spec.selector: Required value": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {},
+				},
+			},
+		},
+		"spec.steps: Invalid value: \"one\": jobTemplate and externalRef are mutually exclusive": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {
+						JobTemplate: &k8sBatch.JobTemplateSpec{},
+						ExternalRef: &k8sApi.ObjectReference{},
+					},
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+	}
+
+	for k, v := range errorCases {
+		errs := ValidateWorkflow(&v)
+		if len(errs) == 0 {
+			t.Errorf("expected failure for %s", k)
+		} else {
+			s := strings.Split(k, ":")
+			err := errs[0]
+			if err.Field != s[0] || !strings.Contains(err.Error(), s[1]) {
+				t.Errorf("unexpected error: %v, expected: %s", err, k)
+			}
+		}
+	}
+}
+
+func NewWorkflow() api.Workflow {
+	return api.Workflow{
+		ObjectMeta: k8sApi.ObjectMeta{
+			Name:            "mydag",
+			Namespace:       k8sApi.NamespaceDefault,
+			UID:             k8sTypes.UID("1uid1cafe"),
+			ResourceVersion: "42",
+		},
+		Spec: api.WorkflowSpec{
+			Steps: map[string]api.WorkflowStep{
+				"one": {
+					JobTemplate: &k8sBatch.JobTemplateSpec{},
+				},
+				"two": {
+					JobTemplate: &k8sBatch.JobTemplateSpec{},
+				},
+			},
+			Selector: &k8sApiUnv.LabelSelector{
+				MatchLabels: map[string]string{"a": "b"},
+			},
+		},
+		Status: api.WorkflowStatus{
+			StartTime: &k8sApiUnv.Time{Time: time.Date(2009, time.January, 1, 27, 6, 25, 0, time.UTC)},
+			Statuses: map[string]api.WorkflowStepStatus{
+				"one": {
+					Complete: false,
+				},
+				"two": {
+					Complete: false,
+				},
+			},
+		},
+	}
+}
+
+func TestValidateWorkflowUpdate(t *testing.T) {
+
+	type WorkflowPair struct {
+		current      api.Workflow
+		patchCurrent func(*api.Workflow)
+		update       api.Workflow
+		patchUpdate  func(*api.Workflow)
+	}
+	errorCases := map[string]WorkflowPair{
+		"metadata.resourceVersion: Invalid value: \"\": must be specified for an update": {
```

Comment:
I don't think we can actually check this using the validator.
I do like throwing an error when no resourceVersion is specified in the client, but I think this is already built in. Maybe create an issue, to research this and implement the error checking if it is not present.